src/libraries/System.Private.CoreLib/src/System/Half.cs

```diff
@@ -0,0 +1,694 @@
+            {
+                if (sig != 0) // NaN
+                {
+                    return CreateHalfNaN(sign, sig << 12); // Shift the significand bits to the left end
+                }
+                return sign ? NegativeInfinity : PositiveInfinity;
+            }
+
+            uint sigHalf = (uint)ShiftRightJam(sig, 38);
+            if ((exp | (int)sigHalf) == 0)
+            {
+                return new Half(sign, 0, 0);
+            }
+            return new Half(RoundPackToHalf(sign, (short)(exp - 0x3F1), (ushort)(sigHalf | 0x4000)));
+        }
+
+        // -----------------------Start of from-half conversions-------------------------
+        public static explicit operator float(Half value)
+        {
+            bool sign = IsNegative(value);
+            int exp = value.Exponent;
+            uint sig = value.Significand;
+
+            if (exp == MaxExponent)
+            {
+                if (sig != 0)
+                {
+                    return CreateSingleNaN(sign, (ulong)sig << 54);
+                }
+                return sign ? float.NegativeInfinity : float.PositiveInfinity;
+            }
+
+            if (exp == 0)
+            {
+                if (sig == 0)
+                {
+                    return CreateSingle(sign ? float.SignMask : 0); // Positive / Negative zero
+                }
+                (exp, sig) = NormSubnormalF16Sig(sig);
+                exp -= 1;
+            }
+
+            return CreateSingle(sign, (byte)(exp + 0x70), sig << 13);
+        }
+
+        public static explicit operator double(Half value)
+        {
+            bool sign = IsNegative(value);
+            int exp = value.Exponent;
+            uint sig = value.Significand;
+
+            if (exp == MaxExponent)
+            {
+                if (sig != 0)
+                {
+                    return CreateDoubleNaN(sign, (ulong)sig << 54);
+                }
+                return sign ? double.NegativeInfinity : double.PositiveInfinity;
+            }
+
+            if (exp == 0)
+            {
+                if (sig == 0)
+                {
+                    return CreateDouble(sign ? double.SignMask : 0); // Positive / Negative zero
+                }
+                (exp, sig) = NormSubnormalF16Sig(sig);
+                exp -= 1;
+            }
+
+            return CreateDouble(sign, (ushort)(exp + 0x3F0), (ulong)sig << 42);
+        }
+
+        // IEEE 754 specifies NaNs to be propagated
+        internal static Half Negate(Half value)
+        {
+            return IsNaN(value) ? value : new Half((ushort)(value._value ^ SignMask));
+        }
+
+        private static (int Exp, uint Sig) NormSubnormalF16Sig(uint sig)
+        {
+            int shiftDist = BitOperations.LeadingZeroCount(sig) - 16 - 5;
+            return (1 - shiftDist, sig << shiftDist);
+        }
+
+        #region Utilities
+
+        private static uint SingleToUInt32Bits(float value)
```

Comment:
nit: I wonder if it would be better to also have these in `BitConverter` or to just put the necessary `signed` to `unsigned` cast at the usage site.

The benefit of the former is that we can more easily track how many places are using the unsigned type which would make it easier to justify making it public.

Reply:
(These also meaning `UInt16BitsToHalf` and `DoubleToUInt64Bits`).

Reply:
If we were to make these methods on BitConverter public, I expect we would expose them as signed - for consistency with the existing similar methods.

Reply:
> I expect we would expose them as signed - for consistency with the existing similar methods.

I agree for `Half`, but my comment was around these helper functions which are equivalent to already exposed methods but returning `unsigned`. i.e. This `SingleToUInt32Bits` as compared to the already public `SingleToInt32Bits`.

I think the simplest thing would be to just use the existing public method and cast at the callsite. But was commenting there may be benefit to having explicit internal helpers to help simplify readability and track how many places actually need unsigned over signed.

Reply:
Done